platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java

```diff
@@ -990,13 +1019,68 @@ public Stream<VirtualHashRecord> dirtyHashesForFlush(final long lastLeafPath) {
     }
 
     /**
-     * Gets estimated number of dirty internal nodes in this cache.
+     * Gets estimated number of dirty hashes in this cache.
+     *
+     * <p>This method and similar methods about leaves are called multiple times. First,
+     * the method is used to estimate this cache copy size to make a decision whether to
+     * flush the copy to disk or to merge it to a newer copy. If the copy is selected to
+     * flush, its {@link #prepareForFlush()} method is called, it may mark some mutations
+     * as filtered, which affects the estimated size. After that this method is called
+     * again, it takes filtered mutations into consideration. If the updated estimated
+     * size is lower than flush threshold, the copy is not flushed to disk, but rather
+     * its {@link #garbageCollect(long, long)} is called.
      *
      * @return
-     *        Estimated number of dirty internal nodes
+     *        Estimated number of dirty hashes
+     */
+    public long estimatedHashesCount() {
+        return (dirtyHashes == null) ? 0 : (dirtyHashes.size() - filteredHashesCount.get());
+    }
+
+    /**
+     * This cache copy is selected to flush to disk, either because it is explicitly
+     * marked as such, or when cache copy estimated size exceeds flush threshold. Before
+     * flush, this method is used to mark redundant mutations as "filtered", so they aren't
+     * included to the streams for the data source. It may happen that after filtering the
+     * cache copy is no longer needed to flush. In this case, all filtered mutations are
+     * removed using {@link #garbageCollect(long, long)} method.
+     */
+    public void prepareForFlush() {
+        // Mark obsolete mutations to filter later and update "filtered" counters. These
+        // counters will affect the estimated size
+        final long version = getFastCopyVersion();
+        filterMutations(dirtyHashes, pathToDirtyHashIndex, version, filteredHashesCount);
+        filterMutations(dirtyLeafPaths, pathToDirtyLeafIndex, version, filteredLeafPathsCount);
+        filterMutations(dirtyLeaves, keyToDirtyLeafIndex, version, filteredLeavesCount);
+    }
+
+    /**
+     * If a cache copy's estimated size exceeded flush threshold, but after mutations are
+     * filtered the size drops below the threshold, the copy is not flushed to disk, but
+     * just removes all filtered mutations using this method.
      */
-    public long estimatedInternalsCount(final long firstLeafPath) {
-        return (dirtyHashes == null) ? 0 : dirtyHashes.size();
+    public void garbageCollect(final long firstLeafPath, final long lastLeafPath) {
+        final Stream<Mutation<Long, Hash>> filteredHashes = dirtyHashes.stream()
+                .filter(mutation -> mutation.key <= lastLeafPath)
+                .filter(mutation -> !mutation.isFiltered())
+                .peek(mutation -> mutation.next = null);
+        dirtyHashes = new ConcurrentArray<>(filteredHashes);
```

Comment:
Minor: this is a public method w/o any concurrency contract defined in its specification, and it modifies a field of this class which may cause issues if this field is accessed from other methods, potentially on other threads. Should the method be private? Should it define its threading model? Should the access to this field be protected/synchronized by any means?

Same comment applies to a few other similar assignments below in this same method.

Reply:
These are all good questions. VirtualNodeCache has a number of methods that are supposed to be called on the virtual lifecycle thread only. Examples are dirtyLeavesForFlush(), release(), and others. Even more important is the phase of the cache. When a cache is just created, it accepts leaf changes, but hashes cannot be modified. Then cache.prepareForHashing() is called, and it switches the gears: now leaf changes are prohibited, but hashes can be updated. Finally, a cache is sealed using seal() method, and from this point on no changes are allowed. In this particular case, the cache must be sealed. I added the corresponding check to the method, as well as a note to the JavaDoc. It will make sure no changes to dirtyHashes / dirtyLeaves / dirtyLeafPaths are done in parallel.